app/models/project.rb

```diff
@@ -322,6 +318,17 @@ def self.visible_by(user = nil)
     end
   end
 
+  # Returns a ActiveRecord::Relation to find all projects for which +user+ has the given +permission+
+  #
+  # Valid options:
+  # * :project => limit the condition to project
+  # * :with_subprojects => limit the condition to project and its subprojects
+  # * :member => limit the condition to the user projects
+  def self.allowed_to(user, permission, options = {})
+    where(allowed_to_condition(user, permission, options))
+      .references(:projects)
```

Comment:
I can understand (barely) the need for `references` on the other models when using a string base condition. But that definitely shouldn't be necessary when being in the Project model already.